orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -1,581 +1,720 @@
-  if (p.size() > s.size()) return false;
-  return std::equal(p.rbegin(), p.rend(), s.rbegin());
+  return Status::OK();
 }
 
-void WriteTensorProtoToFile(const PathString& file_path,
-                            const std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
-                            std::string caller_context) {
-  auto file_write_status = WithOpenFile(
-      file_path, false,
-      [&tensor_protos](int fd) {
-        google::protobuf::io::FileOutputStream output{fd};
-        ORT_RETURN_IF_ERROR(WriteProtoMessageSequence(tensor_protos, output));
-        return Status::OK();
-      });
-
-  ORT_ENFORCE(file_write_status.IsOK(), caller_context, " write file failed: ", ToUTF8String(file_path));
-}
+/**
+ * @brief Create OrtValue objects from flatbuffer tensors.
+ *
+ * @param flatbuffer_tensors Flatbuffer tensors.
+ * @param name_to_ort_value Name to OrtValue map to be populated.
+ * @return Status of the operation.
+ */
+Status OrtValuesFromFlatbufferTensors(
+    const flatbuffers::Vector<flatbuffers::Offset<onnxruntime::fbs::Tensor>>& flatbuffer_tensors,
+    InlinedHashMap<std::string, OrtValue>& name_to_ort_value) {
+  for (auto* fbs_tensor : flatbuffer_tensors) {
+    ORT_RETURN_IF_NOT(fbs_tensor, "Encountered a nullptr flatbuffer tensor. Checkpoint file is invalid.");
+
+    std::string tensor_name;
+    OrtValue ort_value;
+    ORT_RETURN_IF_ERROR(OrtValueFromFlatbufferTensor(*fbs_tensor, tensor_name, ort_value));
+    name_to_ort_value.emplace(tensor_name, ort_value);
+  }
 
-void LoadTensorProtoFromFile(const PathString& file_path,
-                             std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
-                             std::string caller_context) {
-  auto file_read_status = WithOpenFile(
-      file_path, true,
-      [&tensor_protos](int fd) {
-        google::protobuf::io::FileInputStream input{fd};
-        ORT_RETURN_IF_ERROR(ReadProtoMessageSequence(tensor_protos, input));
-        return Status::OK();
-      });
-
-  ORT_ENFORCE(file_read_status.IsOK(), caller_context, " load file failed: ", ToUTF8String(file_path));
+  return Status::OK();
 }
 
-template <typename Func>
-void FilterFilesFromDirectory(const PathString& folder_path, Func func) {
-  LoopDir(folder_path, [&func](const PathChar* filename, OrtFileType file_type) -> bool {
-    if (filename[0] == '.' || file_type == OrtFileType::TYPE_DIR) {
-      return true;
-    }
+namespace Save {
 
-    return func(filename);
-  });
+/**
+ * @brief Save from a checkpoint flatbuffer to file.
+ * @param checkpoint_path Path to save the checkpoint file.
+ * @param builder Flatbuffer builder containing the checkpoint buffer.
+ * @return Status of the operation.
+ *
+ */
+Status ToFile(const PathString& checkpoint_path, flatbuffers::FlatBufferBuilder& builder) {
+  std::ofstream file(checkpoint_path, std::ios::binary);
+  uint8_t* buf = builder.GetBufferPointer();
+  int size = builder.GetSize();
```

Comment:
int32 max is ~ 2GB. Flatbuffer has a file size limit of 2GB I believe (similar to a protobuf message).

We will have to find an alternative in case the checkpoint is that big. My guess though is that a checkpoint that big is not viable for on-device training.